flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/BigQuerySchemaProviderImpl.java

```diff
@@ -0,0 +1,347 @@
+     *
+     * @param tableSchema A {@link TableSchema} object to cast to {@link Schema}.
+     * @return Converted Avro Schema
+     */
+    private static Schema getAvroSchema(TableSchema tableSchema) {
+        return SchemaTransform.toGenericAvroSchema("root", tableSchema.getFields());
+    }
+
+    // --------------- Obtain Descriptor Proto from Avro Schema  ---------------
+    /**
+     * Obtains a Descriptor Proto by obtaining Descriptor Proto field by field.
+     *
+     * <p>Iterates over Avro Schema to obtain FieldDescriptorProto for it.
+     *
+     * @param schema Avro Schema, for which descriptor is needed.
+     * @return DescriptorProto describing the Schema.
+     */
+    private static DescriptorProto getDescriptorSchemaFromAvroSchema(Schema schema) {
+        // Iterate over each table field and add them to the schema.
+        Preconditions.checkState(!schema.getFields().isEmpty());
+
+        DescriptorProto.Builder descriptorBuilder = DescriptorProto.newBuilder();
+        // Create a unique name for the descriptor ('-' characters cannot be used).
+        // Replace with "_" and prepend "D".
+        descriptorBuilder.setName(BigQueryUtils.bqSanitizedRandomUUIDForDescriptor());
+        int i = 1;
+        for (Schema.Field field : schema.getFields()) {
+            fieldDescriptorFromSchemaField(field, i++, descriptorBuilder);
+        }
+        return descriptorBuilder.build();
+    }
+
+    /**
+     * Function to obtain the FieldDescriptorProto from a AvroSchemaField and then append it to
+     * DescriptorProto builder.
+     *
+     * @param field {@link Schema.Field} object to obtain the {@link FieldDescriptorProto} from.
+     * @param fieldNumber index at which the obtained {@link FieldDescriptorProto} is appended in
+     *     the Descriptor.
+     * @param descriptorProtoBuilder {@link DescriptorProto.Builder} object to add the obtained
+     *     {@link FieldDescriptorProto} to.
+     */
+    private static void fieldDescriptorFromSchemaField(
+            Schema.Field field, int fieldNumber, DescriptorProto.Builder descriptorProtoBuilder) {
+
+        @Nullable Schema schema = field.schema();
+        Preconditions.checkNotNull(schema, "Unexpected null schema!");
+
+        FieldDescriptorProto.Builder fieldDescriptorBuilder = FieldDescriptorProto.newBuilder();
+        fieldDescriptorBuilder = fieldDescriptorBuilder.setName(field.name().toLowerCase());
+        fieldDescriptorBuilder = fieldDescriptorBuilder.setNumber(fieldNumber);
+
+        Schema elementType = schema;
+        boolean isNullable = false;
+
+        switch (schema.getType()) {
+            case RECORD:
+                Preconditions.checkState(!schema.getFields().isEmpty());
+                // Check if this is right.
+                DescriptorProto nested = getDescriptorSchemaFromAvroSchema(schema);
+                descriptorProtoBuilder.addNestedType(nested);
+                fieldDescriptorBuilder =
```

Comment:
i'm not 100% sure but i'm pretty sure you dont have to capture the return value when you call set on a builder

Reply:
You are right. Fixed.

Reply:
thanks! you can resolve this convo